model/command_webhook.go

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package model
+
+type CommandWebhook struct {
+	Id        string
+	CreateAt  int64
+	CommandId string
+	UserId    string
+	ChannelId string
+	RootId    string
+	ParentId  string
+	UseCount  int
+}
+
+const (
+	COMMAND_WEBHOOK_LIFETIME = 1000 * 60 * 30
+)
+
+func (o *CommandWebhook) PreSave() {
+	if o.Id == "" {
+		o.Id = NewId()
+	}
+
+	if o.CreateAt == 0 {
+		o.CreateAt = GetMillis()
+	}
+}
+
+func (o *CommandWebhook) IsValid() *AppError {
+	if len(o.Id) != 26 {
+		return NewLocAppError("CommandWebhook.IsValid", "model.command_hook.id.app_error", nil, "")
```

Comment:
I definitely agree with the sentiments behind both of those reasons. To quote the Zen of Python, "There should be one-- and preferably only one --obvious way to do it.". Only having one constructor would be nicer. Likewise, no one ever said consistency was a bad thing.

Let me just try to explain my reasons for thinking `NewAppError` here with the `http.StatusBadRequest` code is bad. Specifically, I'm all for using `NewAppError`, but I think using `http.StatusBadRequest` is bad.

In this case, the input to this function is not defined by the user's request. If one of these errors actually occurs, it's because a codepath was taken that we developers didn't predict such as a race condition or i/o error, and _not_ because the request was malformed.

The "400 Bad Request" status means "The request could not be understood by the server due to malformed syntax.". But we're beyond interpreting the request. We wouldn't hit this point in code if we couldn't understand what the client requested. The 400 status also means "The client SHOULD NOT repeat the request without modifications.". Do we really want to send that signal if an unexpected error happened? When these errors are hit, the client's request is very possibly perfectly fine and should just be attempted again. On the other hand, "500 Internal Server Error" means "The server encountered an unexpected condition which prevented it from fulfilling the request." And that's exactly what this is: an "unexpected condition".

From a maintainer's perspective, masking server errors like this as client errors is a form of [error hiding](https://en.wikipedia.org/wiki/Error_hiding). When you're careful to use 500 vs. 400 errors appropriately, you have much more insight into issues that you need to investigate: Every 500 error should be investigated and fixed. 400 errors can for the most part be ignored.

At it's core, my problem is really that our models' `IsValid` methods are well-informed enough to say what's valid and what isn't, but they are _not_ well-informed enough to assign blame when fields aren't valid. By attempting to assign blame, these methods are making guesses at where their input is originating from. It's a hidden backwards dependency.

I think a much better pattern would be for the `IsValid` methods to _not_ determine a status code. Personally, I like the type of error propagation that packages like [pkg/errors](https://godoc.org/github.com/pkg/errors) can provide:

```go
if err := hook.IsValid(); err != nil {
    return nil, errors.Wrap(BadRequestError, err)
}
```

This says that "there's a bad request error _caused by_ whatever error prevents this hook from being valid", which makes `IsValid` more versatile, while eliminating that hidden dependency by allowing a function that's higher up in the call stack and well-informed enough assign blame.

/rant

Obviously the kind of change I would suggest is beyond the scope of this review. It we really want incorrect, but consistent error codes, I'll make this change when we're closer to being able to merge it.